libs/langchain-community/src/vectorstores/turbopuffer.ts

```diff
@@ -0,0 +1,213 @@
+  apiUrl?: string;
+}
+
+export interface TurbopufferQueryResult {
+  dist: number;
+  id: number;
+  vector: number[];
+  attributes: TurbopufferFilterType;
+}
+
+export class TurbopufferVectorStore extends VectorStore {
+  declare FilterType: TurbopufferFilterType;
+
+  get lc_secrets(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  get lc_aliases(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  // Handle minification for tracing
+  static lc_name(): string {
+    return "TurbopufferVectorStore";
+  }
+
+  protected distanceMetric: TurbopufferDistanceMetric = "cosine_distance";
+
+  protected apiKey: string;
+
+  protected namespace = "default";
+
+  protected apiUrl = "https://api.turbopuffer.com/v1/";
+
+  public _vectorstoreType(): string {
+    return "turbopuffer";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: TurbopufferParams) {
+    super(embeddings, args);
+
+    const apiKey = args.apiKey ?? getEnvironmentVariable("TURBOPUFFER_API_KEY");
+    if (!apiKey) {
+      throw new Error(
+        [
+          "Turbopuffer API key not found.",
+          `Please pass it in as "apiKey" or set it as an environment variable called "TURBOPUFFER_API_KEY"`,
+        ].join("\n")
+      );
+    }
+    this.apiKey = apiKey;
+    this.namespace = args.namespace ?? this.namespace;
+    this.distanceMetric = args.distanceMetric ?? this.distanceMetric;
+    this.apiUrl = args.apiUrl ?? this.apiUrl;
+  }
+
+  getJsonHeader(): TurbopufferHeaders {
+    return {
+      headers: {
+        Authorization: `Bearer ${this.apiKey}`,
+        "Content-Type": "application/json",
+      },
+    };
+  }
+
+  async addVectors(
+    vectors: number[][],
+    documents: DocumentInterface[],
+    options?: { ids?: number[] }
+  ): Promise<void> {
+    try {
+      if (options?.ids && options.ids.length !== vectors.length) {
+        throw new Error(
+          "Number of ids provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length !== vectors.length) {
+        throw new Error(
+          "Number of documents provided does not match number of vectors"
+        );
+      }
+
+      if (documents.length === 0) {
+        throw new Error("No documents provided");
+      }
+
+      const docIds = options?.ids ?? documents.map((_, index) => index);
+
+      const attributes = {
+        source: documents.map((doc) => doc.metadata.source),
+        pageContent: documents.map((doc) => doc.pageContent),
+      };
+
+      const data = {
+        docIds,
+        vectors,
+        attributes,
+      };
+
+      await fetch(`${this.apiUrl}/vectors/${this.namespace}`, {
+        method: "POST",
+        headers: this.getJsonHeader().headers,
+        body: JSON.stringify(data),
+      });
+    } catch (error) {
+      console.error("Error storing vectors:", error);
+      throw error;
+    }
+  }
+
+  async addDocuments(
```

Comment:
It is great that Langchain already seems to encourage batch-adding.

Reply:
We can also add concurrency/more intelligent batching insert like this:

https://github.com/langchain-ai/langchainjs/blob/main/libs/langchain-openai/src/embeddings.ts#L175

 Are there any default values that makes sense to you?

Reply:
The bigger the batch the better. Somewhere around 3,000 seems to work for most dimensionalities.